src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.hpp

```diff
@@ -0,0 +1,283 @@
+// Copyright 2020 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file DiscoveryDataBase.hpp
+ *
+ */
+
+#ifndef _FASTDDS_RTPS_DISCOVERY_DATABASE_H_
+#define _FASTDDS_RTPS_DISCOVERY_DATABASE_H_
+
+#include <vector>
+#include <map>
+#include <mutex>  // For std::unique_lock
+#include <shared_mutex>
+
+#include <fastrtps/utils/fixed_size_string.hpp>
+#include <fastdds/rtps/writer/ReaderProxy.h>
+#include <fastdds/rtps/common/CacheChange.h>
+#include <fastrtps/utils/DBQueue.h>
+
+#include "./DiscoveryDataFilter.hpp"
+#include "./DiscoveryParticipantInfo.hpp"
+#include "./DiscoveryEndpointInfo.hpp"
+#include "./DiscoveryDataQueueInfo.hpp"
+
+namespace eprosima {
+namespace fastdds {
+namespace rtps {
+namespace ddb {
+
+//typedef std::shared_timed_mutex share_mutex_t;
+// only working in C++14
+typedef std::shared_timed_mutex share_mutex_t;
+
+struct CacheChangeCmp
+{
+    bool operator ()(
+            const eprosima::fastrtps::rtps::CacheChange_t& a,
+            const eprosima::fastrtps::rtps::CacheChange_t& b) const
+    {
+        return a.write_params.sample_identity() < b.write_params.sample_identity();
+    }
+
+};
+
+/**
+ * Class to manage the discovery data base
+ *@ingroup DISCOVERY_MODULE
+ */
+class DiscoveryDataBase
+    : public eprosima::fastdds::rtps::ddb::PDPDataFilter<DiscoveryDataBase>
+    , public eprosima::fastdds::rtps::ddb::EDPDataFilter<DiscoveryDataBase>
+    , public eprosima::fastdds::rtps::ddb::EDPDataFilter<DiscoveryDataBase, false>
+{
+
+public:
+
+
+    class AckedFunctor {
+        using argument_type = eprosima::fastrtps::rtps::ReaderProxy*;
+        using result_type = void;
+    public:
+
+        AckedFunctor(DiscoveryDataBase* db, eprosima::fastrtps::rtps::CacheChange_t* change);
+
+        ~AckedFunctor();
+
+        void operator() (eprosima::fastrtps::rtps::ReaderProxy* reader_proxy);
+
+        bool pending()
+        {
+            return pending_;
+        }
+
+    private:
+
+        DiscoveryDataBase* db_;
+        eprosima::fastrtps::rtps::CacheChange_t* cache_;
```

Comment:
I think `change` is clearer. Update in cpp.
```suggestion
        eprosima::fastrtps::rtps::CacheChange_t* change_;
```